linking issue on Solaris - ld: fatal: unrecognized option --strip-debug

While building Rust `master` branch on Solaris with new `beta` compiler from `2024-02-04` (https://github.com/rust-lang/rust/commit/9784a14859aeffa2a3819566ccc77a28f9ae03be) I get following error:

```
Building bootstrap
   Compiling libc v0.2.151
..
   Compiling termcolor v1.4.0
   Compiling home v0.5.9
error: linking with `gcc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/bin:/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/bin:/usr/gnu/bin/:/usr/bin" VSLANG="1033" "gcc" "-m64" "/tmp/rustcVudrFx/symbols.o" "/builds/psumbera/rust-lang-build/build/bootstrap/debug/build/serde_json-3d0a0f4e969b1971/build_script_build-3d0a0f4e969b1971.build_script_build.22f20d396b540307-cgu.0.rcgu.o" "/builds/psumbera/rust-lang-build/build/bootstrap/debug/build/serde_json-3d0a0f4e969b1971/build_script_build-3d0a0f4e969b1971.2nfbgxrvkxklbize.rcgu.o" "-Wl,-z,ignore" "-L" "/builds/psumbera/rust-lang-build/build/bootstrap/debug/deps" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib" "-Wl,-Bstatic" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libstd-9928786342ee8f0e.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libpanic_unwind-82e2fefe18f026cf.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libobject-c0b7b4100ea52d7a.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libmemchr-9f92b5e161242b5c.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libaddr2line-c97e97ed47cabe77.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libgimli-3ee598c57fa4a8fe.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/librustc_demangle-b9f4264552f8e9dc.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libstd_detect-b1f18a4f4581f203.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libhashbrown-167eeefb37a594f2.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/librustc_std_workspace_alloc-e69a34bb5b469969.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libminiz_oxide-e8ecd3fd30775672.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libadler-bb51104b5c71d37c.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libunwind-929c7e8ad513a12d.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libcfg_if-0539aff4cd5dd71d.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/liblibc-7f1958c646d5977f.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/liballoc-bd0653933ca12878.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/librustc_std_workspace_core-8173e93f72ebef81.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libcore-e825afa3781f99a6.rlib" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib/libcompiler_builtins-a913ffd08d993ff1.rlib" "-Wl,-Bdynamic" "-lsocket" "-lposix4" "-lpthread" "-lresolv" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lsendfile" "-llgrp" "-L" "/builds/psumbera/rust-lang-build/build/x86_64-pc-solaris/stage0/lib/rustlib/x86_64-pc-solaris/lib" "-o" "/builds/psumbera/rust-lang-build/build/bootstrap/debug/build/serde_json-3d0a0f4e969b1971/build_script_build-3d0a0f4e969b1971" "-Wl,--strip-debug" "-nodefaultlibs"
  = note: ld: fatal: unrecognized option --strip-debug
          ld: fatal: use the '-z help' option for usage information
          collect2: error: ld returned 1 exit status
```

I was trying to bisect the issue in `beta` and it reports the issue comes from `cargo` update here https://github.com/rust-lang/rust/commit/7e1e61a4478b690113c86bf7b547e829969fe29d.

I haven't tried to bisect it further but it seems that it comes from this particuar `cargo` change https://github.com/rust-lang/cargo/commit/bc534513dd1cef1625ff30df733141971570e531 (https://github.com/rust-lang/cargo/pull/13257).

Fix or workaround seems to be to apply similar changes as Illumos did some time ago here https://github.com/rust-lang/rust/commit/f3deac2559e2354599db4c7a1c546fedd288ba9f. Particularly here: https://github.com/rust-lang/rust/blob/0987e41d1ccbae7a8a55fdcde4af0ea76ebfe2d3/compiler/rustc_codegen_ssa/src/back/linker.rs#L629-L631

--- a/compiler/rustc_codegen_ssa/src/back/linker.rs
+++ b/compiler/rustc_codegen_ssa/src/back/linker.rs
@@ -626,7 +626,7 @@ impl<'a> Linker for GccLinker<'a> {
                 // it does support --strip-all as a compatibility alias for -s.
                 // The --strip-debug case is handled by running an external
                 // `strip` utility as a separate step after linking.
-                if self.sess.target.os != "illumos" {
+                if !self.sess.target.is_like_solaris {
                     self.linker_arg("--strip-debug");
                 }
             }

